### PR TITLE
Develop button to create role

### PR DESCRIPTION
**Pull request by plinionr3**
The creation of a button that allows a user to create a new role.

### CODE REVIEW

File: backend/src/main/java/saci/application/RoleController.java [hunk 1]
```diff
@@ -40,8 +42,9 @@ public class RoleController {
             })
     @ResponseStatus(HttpStatus.CREATED)
     @PostMapping
-    public Role createRole(@RequestBody Role role) {
-        return roleService.createRole(role);
+    public ResponseEntity<Role> createRole(@RequestBody Role role) {
+        Role createdRole = roleService.createRole(role);
+        return createdRole == null ?   ResponseEntity.status(HttpStatus.CONFLICT).build() : ResponseEntity.ok(createdRole);
```
Comment (vitorlofonseca) [inline backend/src/main/java/saci/application/RoleController.java:47]: Run in the backend folder `mvn spotless:apply` to take out these empty spaces in the code (Spotless also format the code, like Prettier in frontend)

File: backend/src/main/java/saci/application/RoleController.java [hunk 2]
```diff
@@ -40,8 +42,9 @@ public class RoleController {
             })
     @ResponseStatus(HttpStatus.CREATED)
     @PostMapping
-    public Role createRole(@RequestBody Role role) {
-        return roleService.createRole(role);
+    public ResponseEntity<Role> createRole(@RequestBody Role role) {
```
Comment (vitorlofonseca) [inline backend/src/main/java/saci/application/RoleController.java:45]: Have you verified if role object has the name attribute? 

Tip: you can use @Valid annotation

File: backend/src/main/java/saci/domain/service/RoleService.java [hunk 3]
```diff
@@ -12,8 +14,10 @@ public class RoleService {
 
     private RoleRepository roleRepository;
 
+
     public Role createRole(Role role) {
-        return roleRepository.save(role);
+        Optional<Role> optionalRole = roleRepository.findByName(role.getName());
+        return optionalRole.isPresent() ? null : roleRepository.save(role);
```
Comment (vitorlofonseca) [inline backend/src/main/java/saci/domain/service/RoleService.java:20]: Instead of return null here, take a look how you could use this exception

https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/web/client/HttpClientErrorException.Conflict.html

If you don't know how to throw exceptions in Java, check this
https://rollbar.com/guides/java/how-to-throw-exceptions-in-java/

File: frontend/src/stores/roles/actions.ts [hunk 4]
```diff
@@ -13,9 +13,9 @@ async function fetchRoles(): Promise<void> {
   setRoles(roles)
 }
 
-async function addRole(role: Role): Promise<void> {
-  await post<Role[]>('/roles', role)
+async function addRole(role: Role): Promise<Role[]> {
```
Comment (vitorlofonseca) [inline frontend/src/stores/roles/actions.ts:16]: We don't need this return anymore, since we are using the exception. Can you revert the modifications in this file?

File: frontend/src/views/DemoSection.vue [hunk 5]
```diff
@@ -17,6 +17,7 @@ const onSaveRole = async () => {
   ElMessage({
     message: 'Congrats, your role is now added',
     type: 'success'
+
```
Comment (vitorlofonseca) [inline frontend/src/views/DemoSection.vue:20]: Run `yarn lint` in the frontend project, to take out these spaces

File: frontend/src/views/HomeView.vue [hunk 6]
```diff
@@ -1 +1,89 @@
-<template><h1>Test</h1></template>
+<script lang="ts" setup>
+import { onMounted, ref, reactive } from 'vue'
+import { ElButton, ElDialog, ElForm, ElFormItem, ElInput, ElMessage } from 'element-plus'
+import { useRolesStore } from '@/stores/roles/index'
+import DemoSection from './DemoSection.vue'
+
+const dialogFormVisible = ref(false)
+const formLabelWidth = '100px'
```
Comment (vitorlofonseca) [inline frontend/src/views/HomeView.vue:8]: If you are not using this variable, please remove it

File: frontend/src/views/HomeView.vue [hunk 7]
```diff
@@ -1 +1,89 @@
...
+      <ElInput v-model="newRoleName" placeholder="Role name" :clearable="false" />
+
+      <template #footer>
+        <div class="dialog-footer">
+          <ElButton @click="dialogFormVisible = false">Cancel</ElButton>
+          <ElButton type="primary" @click="createRole" :required="true"> Confirm </ElButton>
+        </div>
+      </template>
+    </ElDialog>
+    <DemoSection v-if="activeView === Views.demo" />
```
Comment (vitorlofonseca) [inline frontend/src/views/HomeView.vue:71]: Why are you using DemoSection component? Can you remove it?

File: frontend/src/views/HomeView.vue [hunk 8]
```diff
@@ -1 +1,89 @@
...
+      }
+    }
+  }
+
+  newRoleName.value = ''
+}
+</script>
+
+<template>
+  <div class="demo">
```
Comment (vitorlofonseca) [inline frontend/src/views/HomeView.vue:59]: This class is not related to demo. Can you rename it?

File: frontend/src/views/HomeView.vue [hunk 9]
```diff
@@ -1 +1,89 @@
...
+    <ElButton plain @click="dialogFormVisible = true"> Add new Role </ElButton>
```
Comment (vitorlofonseca) [inline frontend/src/views/HomeView.vue:60]: Can you put this click event in a function inside script tag?

File: frontend/src/views/HomeView.vue [hunk 10]
```diff
@@ -1 +1,89 @@
...
+    <ElDialog v-model="dialogFormVisible" title="New Role" width="500">
+      <ElInput v-model="newRoleName" placeholder="Role name" :clearable="false" />
+
+      <template #footer>
+        <div class="dialog-footer">
+          <ElButton @click="dialogFormVisible = false">Cancel</ElButton>
```
Comment (vitorlofonseca) [inline frontend/src/views/HomeView.vue:66]: Same here. Please, avoid to put logic inside template tag, always inside script tag

File: frontend/src/views/HomeView.vue [hunk 11]
```diff
@@ -1 +1,89 @@
...
+          <ElButton type="primary" @click="createRole" :required="true"> Confirm </ElButton>
```
Comment (vitorlofonseca) [inline frontend/src/views/HomeView.vue:67]: Why this `:required`?

File: frontend/src/views/HomeView.vue [hunk 12]
```diff
@@ -1 +1,89 @@
...
+        </div>
+      </template>
+    </ElDialog>
+    <DemoSection v-if="activeView === Views.demo" />
+  </div>
+</template>
+
+<style scoped lang="scss">
+.demo {
+  min-height: 100vh;
```
Comment (vitorlofonseca) [inline frontend/src/views/HomeView.vue:77]: Why `min-height`, and not `height`?

File: frontend/src/views/HomeView.vue [hunk 13]
```diff
@@ -1 +1,89 @@
...
+  display: flex;
+  flex-direction: column;
+  justify-content: center;
+  align-items: center;
+  gap: 20px;
+
+  .el-input,
```
Comment (vitorlofonseca) [inline frontend/src/views/HomeView.vue:84]: There is a `components.scss` file, where you can put these generic styles (related to components). Can you put these there?

File: frontend/src/views/HomeView.vue [hunk 14]
```diff
@@ -1 +1,89 @@
...
+  await rolesStore.fetchRoles()
+})
+
+const createRole = async () => {
+  if (newRoleName.value == '') {
+    ElMessage({
+      message: 'you need to fill in this field',
+      type: 'error'
+    })
+  } else {
```
Comment (vitorlofonseca) [inline frontend/src/views/HomeView.vue:29]: If you return in the previous `if` statement, this `else` statement is not necessary

File: frontend/src/views/HomeView.vue [hunk 15]
```diff
@@ -1 +1,89 @@
...
+const activeView = ref(Views.references)
+
+onMounted(async () => {
+  await rolesStore.fetchRoles()
+})
+
+const createRole = async () => {
+  if (newRoleName.value == '') {
+    ElMessage({
+      message: 'you need to fill in this field',
```
Comment (vitorlofonseca) [inline frontend/src/views/HomeView.vue:26]: Capital case in the message, please

File: frontend/src/views/HomeView.vue [hunk 16]
```diff
@@ -1 +1,89 @@
...
+      type: 'error'
+    })
+  } else {
+    dialogFormVisible.value = false
+    try {
+      await rolesStore.addRole({ name: newRoleName.value })
+
+      ElMessage({
```
Comment (vitorlofonseca) [inline frontend/src/views/HomeView.vue:34]: Please, create a new file under `services` folder called `messages`. This file will contain the following functions:

ShowSuccessMessage
ShowErrorMessage

These methods just receive a parameter, a string that is the message itself. This way, you don't need to call ElMessage everytime.

File: frontend/src/views/HomeView.vue [hunk 17]
```diff
@@ -1 +1,89 @@
...
+      })
+    } catch (error: Error) {
+      if (error.status === 409) {
+        ElMessage({
+          message: 'Role name already exists.',
+          type: 'error'
+        })
+      } else {
+        // Handle other status codes
+        ElMessage({
```
Comment (vitorlofonseca) [inline frontend/src/views/HomeView.vue:46]: You could put this generic error message inside http.ts file, where all the requests to the backend pass through

File: frontend/src/views/HomeView.vue [hunk 18]
```diff
@@ -1 +1,89 @@
...
+    dialogFormVisible.value = false
+    try {
+      await rolesStore.addRole({ name: newRoleName.value })
+
+      ElMessage({
+        message: 'Congrats, your role is now added',
+        type: 'success'
+      })
+    } catch (error: Error) {
+      if (error.status === 409) {
```
Comment (vitorlofonseca) [inline frontend/src/views/HomeView.vue:39]: Random numbers in the code are not good. What is the mean of 409?

Please create a file with a constant like `HTTP_STATUS_CODE_CONFLICT = 409`, and use it here

**Review by vitorlofonseca — CHANGES_REQUESTED** [review]
Good job!

**Review by PedroMagalhaes20 — APPROVED** [review]

File: frontend/package-lock.json
Comment (vitorlofonseca) [inline frontend/package-lock.json:1]: Can you remove this file? This file is a result of `npm install` command, and we are using just `yarn install`

File: frontend/src/components/DividerComponent.vue
Comment (vitorlofonseca) [inline frontend/src/components/DividerComponent.vue:1]: /components/DividerComponent is redundant. Can you go back to just Divider?

File: frontend/src/services/http.ts [hunk 19]
```diff
@@ -21,8 +23,13 @@ const doRequest = async <T>(url: string, method: string, body?: object | string)
     body
   })
 
+  const HTTP_SERVER_ERROR = 500
```
Comment (vitorlofonseca) [inline frontend/src/services/http.ts:26]: Can you put this variable outside of this function? This way, you can use this constant in any other function. Also, bring the 409 error, put here, and import there in the view component

File: frontend/src/services/messages/ErrorMessage.ts
Comment (vitorlofonseca) [inline frontend/src/services/messages/ErrorMessage.ts:1]: Since you already have `services/messages.ts`, can you delete these files?

File: frontend/src/stores/roles/actions.ts (hunk 4 above)
Comment (vitorlofonseca) [reply frontend/src/stores/roles/actions.ts:16]: Have you undone the modifications of this file?

File: frontend/src/theme/components.scss [hunk 20]
```diff
@@ -13,3 +13,16 @@ button,
 input {
   font-family: 'Noto Sans';
 }
+.newRole {
```
Comment (vitorlofonseca) [inline frontend/src/theme/components.scss:16]: This class is being used inside HomeView. Can you put there, and just let here the general things, like el-input and h4?

File: frontend/src/views/HomeView.vue [hunk 21]
```diff
@@ -1 +1,58 @@
-<template><h1>Test</h1></template>
+<script lang="ts" setup>
+import { onMounted, ref } from 'vue'
+import { ElButton, ElDialog, ElInput } from 'element-plus'
+import { useRolesStore } from '@/stores/roles/index'
+import { ErrorMessage } from '@/services/messages/messages'
+import { SuccessMessage } from '@/services/messages/messages'
+
+const dialogFormVisible = ref(false)
+const HTTP_STATUS_CODE_CONFLICT = 409
```
Comment (vitorlofonseca) [inline frontend/src/views/HomeView.vue:9]: Put this constant in the http.ts file, please

File: frontend/src/views/HomeView.vue [hunk 22]
```diff
@@ -1 +1,58 @@
...
+  dialogFormVisible.value = true
+}
+
+const onClickCancel = () => {
+  dialogFormVisible.value = false
+}
+
+const createRole = async () => {
+  if (newRoleName.value == '') {
+    ErrorMessage('You need to fill in this field')
```
Comment (vitorlofonseca) [inline frontend/src/views/HomeView.vue:27]: You can return after this message. This way, you won't need the else statement

File: frontend/src/services/http.ts [hunk 23]
```diff
@@ -36,6 +41,11 @@ const doRequest = async <T>(url: string, method: string, body?: object | string)
   return Promise.resolve(result)
 }
 
+enum HttpStatusCode {
```
Comment (vitorlofonseca) [inline frontend/src/services/http.ts:44]: And if some other function needs to use this enum? Now it's just inside of function scope. Can you put it outside this function, in this same file?

File: frontend/src/views/HomeView.vue [hunk 24]
```diff
@@ -1 +1,69 @@
-<template><h1>Test</h1></template>
+<script lang="ts" setup>
+import { onMounted, ref } from 'vue'
+import { ElButton, ElDialog, ElInput } from 'element-plus'
+import { useRolesStore } from '@/stores/roles/index'
+import { ErrorMessage } from '@/services/messages/messages'
```
Comment (vitorlofonseca) [inline frontend/src/views/HomeView.vue:5]: You can merge this import, with the below one. If the import come from the same file, you can put in the same line (as element-plus above)

File: frontend/src/views/HomeView.vue [hunk 25]
```diff
@@ -1 +1,69 @@
...
+import { SuccessMessage } from '@/services/messages/messages'
+import { HttpStatusCode } from '@/services/http'
+
+const dialogFormVisible = ref(false)
+const rolesStore = useRolesStore()
+const newRoleName = ref('')
+
+onMounted(async () => {
+  await rolesStore.fetchRoles()
```
Comment (vitorlofonseca) [inline frontend/src/views/HomeView.vue:14]: Why are you fetching roles here?

File: frontend/src/views/HomeView.vue (hunk 22 above)
Comment (vitorlofonseca) [reply frontend/src/views/HomeView.vue:27]: If you don't return in this case, the function will try to create the role, even showing the message. Have you tested this?

File: frontend/src/views/HomeView.vue [hunk 26]
```diff
@@ -1 +1,69 @@
...
+}
+
+const onClickCancel = () => {
+  dialogFormVisible.value = false
+}
+
+const createRole = async () => {
+  if (newRoleName.value == '') {
+    ErrorMessage('You need to fill in this field')
+    dialogFormVisible.value = true
```
Comment (vitorlofonseca) [inline frontend/src/views/HomeView.vue:28]: You could rename the functions to show/hide the dialog, and use it here also (since it's the same logic)

**Review by vitorlofonseca — APPROVED** [review]
